sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -404,9 +408,96 @@ private JmsIO() {}
     T mapMessage(Message message) throws Exception;
   }
 
+  /**
+   * Checkpoint for an unbounded JMS source. Consists of the JMS messages waiting to be acknowledged
+   * and oldest pending message timestamp.
+   */
+  @VisibleForTesting
+  static class JmsCheckpointMark implements UnboundedSource.CheckpointMark, Serializable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JmsCheckpointMark.class);
+
+    @VisibleForTesting Instant oldestMessageTimestamp = Instant.now();
+    @VisibleForTesting transient List<Message> messages = new ArrayList<>();
```

Comment:
A `CheckpointMark` is basically an object to save pending unprocessed state, in the case of JmsIO pending state are the messages. If we make the messages `transient` we are losing data. So we need to fix this to make the messages part of the serialized state of the object. However since Jms `Message`s are not `Serializable` probably what we need to store here are `JmsRecord`s.

Reply:
Possible issue after changing the pending state to be `JmsRecord`s is the fact that the records need to recreate Messages to be able to ack on `finalizeCheckpoint`. Doing ack on the CheckpointMark is probably an error because you would need also the connection information, so better to move that logick out of the CheckpointMark.

Reply:
Other extra bit is that `JmsRecord` equals and hashcode implementation do not take into account the same fields, this should be also addressed otherwise the equals contract might get broken.

Reply:
That's not possible, simply to execute the ack on messages. So, I'm using the same approach as in MqttIO. Just keep the oldest timestamp, and don't ack the messages. Then, the messages stay in the broker and will be consumed by another executor.

Reply:
The finalize checkpoint should ack the message. A better impl is:
1. use auto_ack on session
2. convert Message as JmsRecord and store in the checkpoint mark
3. thanks to that, the record will be serializable and we can transport it.

I'm doing the change.

Reply:
@jbonofre  Could this logic be implemented using `JmsCheckpointMark.State` class as it was before? Well, for this case, at least, we need to make `State` serialisable but it should not be a problem with conversion from `Message` to `JmsRecord`.

Reply:
I'm also wondering if we can have OOM here if the size of messages is large?

Reply:
We can limit the size of the messages list and call finalize checkpoint. State is useless IMHO, I would keep it in checkpoint mark as it's more straight forward to understand.